File: benchmarks/fixtures/webapp_go/internal/services/session_service.go
package services

import (
    "webapp_go/internal/database"
    "webapp_go/internal/models"
    "webapp_go/pkg/logger"
)

var sessSvcLog = logger.GetLogger("services.session")

// SessionService manages user sessions.
type SessionService struct {
    BaseServiceImpl
    DB *database.DatabaseConnection
}

// NewSessionService creates a new session service.
func NewSessionService(db *database.DatabaseConnection) *SessionService {
    sessSvcLog.Info("Creating SessionService")
    return &SessionService{
        BaseServiceImpl: BaseServiceImpl{
            ServiceName:    "session",
            ServiceVersion: "1.0",
        },
        DB: db,
    }
}

// Create starts a new session.
func (s *SessionService) Create(userID, token, ip, userAgent string) *models.Session {
    sessSvcLog.Info("Creating session for user: %s", userID)
    session := models.NewSession(userID, token, ip, userAgent)
    s.DB.Insert("sessions", map[string]interface{}{
        "user_id": userID,
        "token":   token,
    })
    return session
}

// Invalidate revokes a session.
func (s *SessionService) Invalidate(sessionID string) error {
    sessSvcLog.Info("Invalidating session: %s", sessionID)
    return s.DB.Delete("sessions", sessionID)
}

// InvalidateAll revokes all sessions for a user.
func (s *SessionService) InvalidateAll(userID string) error {
    sessSvcLog.Info("Invalidating all sessions for user: %s", userID)
    _, err := s.DB.ExecuteQuery("DELETE FROM sessions WHERE user_id = $1", userID)
    return err
}

// FindByToken looks up a session by token.
func (s *SessionService) FindByToken(token string) (*models.Session, error) {
    sessSvcLog.Info("Finding session by token")
    results, err := s.DB.ExecuteQuery("SELECT * FROM sessions WHERE token = $1", token)
    if err != nil {
        return nil, err
    }
    if len(results) == 0 {
        return nil, nil
    }
    return &models.Session{Token: token, Status: models.SessionActive}, nil
}

File: benchmarks/fixtures/webapp_c/services/session_service.c
#include <stddef.h>

#include "services/session_service.h"

#include "database/database_connection.h"
#include "errors/errors.h"
#include "services/base_service.h"
#include "util/logger.h"

static void session_service_on_initialize(struct BaseService *self)
{
    struct Logger log = get_logger("services.session");
    logger_debug(&log, self->service_name);
}

static const struct BaseServiceVTable SESSION_SERVICE_VTABLE = {
    session_service_on_initialize,
    NULL,
    NULL
};

struct SessionService session_service_new(struct DatabaseConnection *db)
{
    struct SessionService service;
    service.base = base_service_new("session", &SESSION_SERVICE_VTABLE);
    service.db = db;
    return service;
}

enum AppError session_service_create(struct SessionService *service, const char *token)
{
    struct Logger log = get_logger("services.session");
    logger_info(&log, "Creating session");

    enum AppError err = require_initialized(&service->base);
    if (err != APP_OK) {
        return err;
    }

    insert(service->db, "sessions", "INSERT INTO sessions (token) VALUES (?)");
    (void)token;
    return APP_OK;
}

enum AppError session_service_destroy(struct SessionService *service, const char *token)
{
    struct Logger log = get_logger("services.session");
    logger_info(&log, "Destroying session");
    execute_query(service->db, "DELETE FROM sessions WHERE token = ?");
    (void)token;
    return APP_OK;
}
